### ISSUE TITLE
show warning if delivery date < latest availability date on orders

### ISSUE THREAD

**Issue by kbower** [kbower/tickettest1#1246]
_Imported from trac ticket 1246.
Created by **rneville**
Opened in trac: Nov 21, 2011, 9:04:14 AM
Last modified in trac: Aug 6, 2013, 3:03:49 PM_
***

***
* kbower attached [untitled-part.png](https://trac.retailarchitects.com/trac/attachment/ticket/1246/untitled-part.png) on Jun 20, 2013, 8:05:32 AM
* criley attached [ticket1246.patch](https://trac.retailarchitects.com/trac/attachment/ticket/1246/ticket1246.patch) on Jun 24, 2013, 9:44:52 AM

**Comment by kbower** [comment]
_**rneville** trac comment on Jun 14, 2012, 3:13:52 PM:_

Calvin, come see one of us before beginning on this one.

**Comment by kbower** [comment]
_**kbower** trac comment on Jun 18, 2012, 8:44:48 AM:_

I just implemented almost this exact rule for Kane's except that Kane's wanted to ignore bedding.  The rule needs to come **after** rules are run because that is where we reserve inventory.  

`Order.auditing()` is the appropriate method, but it should invoke another method which then you'll need to move the Kane's logic to in the Kane's rule.

**Comment by kbower** [comment]
_**criley** trac comment on Jun 14, 2013, 2:15:55 PM:_

(Changeset [[4244]](../commit/11104211192a6232756dfbd7b9833afbaa549c39))
* References #1246 (show warning if delivery date < latest availability date on orders)

**Comment by kbower** [comment]
_**criley** trac status change on Jun 14, 2013, 3:37:58 PM:_ **closed**

**Comment by kbower** [comment]
_**kbower** trac comment on Jun 17, 2013, 9:30:26 AM:_

These changes aren't compatible with isf rules: the isf_available_for_site() method signature has been changed.  This requires we search code for all the occurrences of the method and possibly update them.  See isf.py, which function will crash now when invoked.  (Reopening ticket)

Regarding comment:2, does the order atpdate factor out filled lines?  I'm pretty sure this warning check needs to come after the rules since lines may be filled that atpdate does not account for.  

Also, we cannot arbitrarily add 5 days to the atp date by default.  That is very much a business rule determined by the retailer.  I don't think we can let this be released as is: won't it always show atpdate 5 days in the future, even for things in stock?  This can't be behavior we want, right?  (Unless we have checked with Kane's/other retailers and every retailer universally wants 5 days future time, even for in stock items, this needs refactoring.)

**Comment by kbower** [comment]
_**kbower** trac status change on Jun 17, 2013, 9:30:26 AM:_ **reopened**

**Comment by kbower** [comment]
_**rneville** trac comment on Jun 17, 2013, 9:37:05 AM:_

Was the 5 day cushion added as part of this ticket or was that already there?  I'd so should be opened as another ticket.

**Comment by kbower** [comment]
_**criley** trac comment on Jun 18, 2013, 10:09:17 AM:_

(Changeset [[4250]](../commit/cd03834717b4461651d2fbf24dc41c133ec53a0a))
* Fixes #1246 (show warning if delivery date < latest availability date on orders)

**Comment by kbower** [comment]
_**criley** trac status change on Jun 18, 2013, 10:09:17 AM:_ **closed**

**Comment by kbower** [comment]
_**kbower** trac comment on Jun 20, 2013, 8:05:32 AM:_

I haven't communicated comment:2 very well; a picture is better:

[[Image(untitled-part.png)]]

Please see comment:2 and comment:6. Invoking atpdate where it's been implemented is breaking it. Now atpdate is being force evaluated _before_ the rules have run which may fill lines that atpdate calculations (notoriously bad) missed. We end up with `**Unknown**` Availability date and a warning, even though the line is _filled_. `atpdate` used to evaluate during serialization, after everything else was run, including fill rules.

Also, do not use jargon to communicate with users. Loft caters to "gap girl" who shouldn't have to try to guess what "ATP" means. Replace your "ATP" terminology with "availability date" in messages the user will read.

This warning should read something like "Delivery has been set despite an unknown availability date"

**Comment by kbower** [comment]
_**kbower** trac status change on Jun 20, 2013, 8:06:02 AM:_ **reopened**

**Comment by kbower** [comment]
_**kbower** trac comment on Jun 20, 2013, 8:16:55 AM:_

Another concern: the issue with this ticket is its been implemented without choice by the retailer while most our retailer's atpdate rules are faulty. So adding a warning is likely to create warning noise and help users decide warnings aren't important enough to read.

It would be nice if this warning can be enabled/disabled to avoid this (either by rule or system parameter).

**Comment by kbower** [comment]
_**rneville** trac comment on Jun 21, 2013, 12:50:35 PM:_

in light of today's phone call I think the priority of this is higher....

**Comment by kbower** [comment]
_**rneville** trac comment on Jun 24, 2013, 10:41:12 AM:_

Do you have the system parameter referenced added to systemdata.py (deliverydate_atpdate_message_option)?

Also, I realize it makes the code a little more verbose, but I would prefer if the options where more meaningful to the user. So I would make the system parameter like this:
```
('deliverydate_atpdate_message_option',"Handling of delivery scheduling prior to ATP date [(W)arn, (E)rror, (I)gnore]",'C','s','N',None,None,),
```
And make it react that way in your code to a letter code. It also helps with readability since I don't have to spend an extra couple of seconds interpreting <=> 0 when reading it as a programmer.

Otherwise, looks good to me. Change those up and it should be good to go. Thanks!

**Comment by kbower** [comment]
_**criley** trac comment on Jun 24, 2013, 12:00:17 PM:_

(Changeset [[4275]](../commit/e8d40a29490552a8448c5bd73ca44d24178d5512))
* Fixes #1246 (show warning if delivery date < latest availability date on orders)

**Comment by kbower** [comment]
_**criley** trac status change on Jun 24, 2013, 12:00:17 PM:_ **closed**

**Comment by kbower** [comment]
_**kbower** trac comment on Jul 23, 2013, 3:54:48 PM:_

(Changeset [[4366]](../commit/6b11c0f81b6af127e246ce83469f118857b207e9))
* change "Delivery Date set with an unknown available date" warning message, which was confusing users #all
   * new message: "Delivery is scheduled even though inventory availablity date is unknown"
 * while "warning" is the ideal Loft default, until ATP rules have been set up properly, warning for this circumstance just causes more confusion and creates to many false positives (false warnings), so we will make Loft's default "ignore", see #1246

**Comment by kbower** [comment]
_**kbower** trac comment on Aug 6, 2013, 3:03:49 PM:_

(Changeset [[4410]](../commit/f6f78ae590a3ce244fc5dc0d1d6e0b8eb9623225))
* add permission 'orderdeliverybeforeatpdate': "ORDERS - Ability to schedule delivery to date before order availability date." #all
   * this allows 'deliverydate_vs_atpdate' system parameter to be set to [E]rror, while still allowing a manager to make an exception for incorrect atpdate
   * this should be for extreme exceptions only and so the inconvenient caveat is this: _any save to the order, even just adding a comment, will require manager override again_ 
   * fixes [#2392](2392), see #1246
